CONVENTIONS.md

```diff
@@ -143,7 +143,47 @@ Please note that in case of [Single Node OpenShift](https://github.com/openshift
   * Set soft pod anti-affinity on the hostname so that pods prefer to land on separate nodes (will be violated on two node clusters)
   * Use maxSurge for deployments if possible since the spreading rules are soft
 
-In future, when we include the descheduler into OpenShift by default, it will periodically rebalance the cluster to ensure spreading for operand or operator is honored. At that time we will remove hard anti-affinity constraints, and recommend components move to a surge model during upgrades.
+In the future, when we include the descheduler into OpenShift by default, it will periodically rebalance the cluster to ensure spreading for operand or operator is honored. At that time we will remove hard anti-affinity constraints, and recommend components move to a surge model during upgrades.
+
+##### Handling kube-apiserver disruption
+
+kube-apiserver disruption can happen for multiple reasons, including
+1. kube-apiserver rollout on a non-HA cluster
+2. networking disruption on the host running the client
+3. networking disruption on the host running the server
+4. internal load balancer disruption
+5. external load balancer disruption
+
+We have seen all of these cases, and more, disrupt connections.
+Many workloads, controllers, and operators rely on the kube-apiserver for making authentication, authorization, and leader election.
+To avoid disruption and mass pod-death, it is important to 
+1. /healthz, /readyz, /livez should not require authorization.
+   They are already open to unauthenticated, so the delegated authorization check presents a reliability risk without
+   a security benefit.
+   This is now the default in the delegated authorizer in k8s.io/apiserver based servers, I'm unsure of controller-runtime.
+2. Binaries should terminate in-cluster client certificate connection.
+   This allows for authentication without contacting the kube-apiserver.
+   The canonical case here is the metrics scraper.
+   In 4.9, the metrics scraper will support using in-cluster client-certificates to increase reliability of scraping
+   in cases of kube-apiserver disruption.
+   This is now the default in the delegated authenticator in k8s.io/apiserver based servers, I'm unsure of controller-runtime.
```

Comment:
Controller runtime doesn't support TLS for its metrics endpoint, typically you have to deploy kube-rbac-proxy in front of your metrics endpoint to allow a TLS connection for metrics. Kube-RBAC-Proxy talks to the API server a lot, this will cause issues and will need some thought to workaround.

We have had some discussion within CR before about allowing the endpoints, such as metrics endpoints to be extendable, eg to allow a user to implement their own TLS server, or add middlewares, but I don't think anything ever came of it

Reply:
> Controller runtime doesn't support TLS for its metrics endpoint, typically you have to deploy kube-rbac-proxy in front of your metrics endpoint to allow a TLS connection for metrics. Kube-RBAC-Proxy talks to the API server a lot, this will cause issues and will need some thought to workaround.

That's a significant shortcoming that was resolved in kube several years ago.  Someone motivated to use controller-runtime probably ought to try to get it updated or perhaps switch to using the upstream project (k8s.io/client-go, etc).

The kube-rbac-proxy will need to support it in order to be adopted.